tests/test_opds_import.py

```diff
@@ -535,23 +536,69 @@ def test_extract_metadata(self):
         [failure] = failures.values()
         eq_(u"202: I'm working to locate a source for this identifier.", failure.exception)
 
-    def test_use_dcterm_identifier_as_id(self):
+    def test_use_dcterm_identifier_as_id_with_id_and_dcterms_identifier(self):
         data_source_name = "Data source name " + self._str
+        collection_to_test = self._default_collection
+        collection_to_test.primary_identifier_source = ExternalIntegration.DCTERMS_IDENTIFIER
         importer = OPDSImporter(
-            self._db, collection=None, data_source_name=data_source_name
+            self._db, collection=collection_to_test, data_source_name=data_source_name,
         )
+
         metadata, failures = importer.extract_feed_data(
-            self.feed_with_id_and_dcterms_identifier,
-            custom_identifier=ExternalIntegration.DCTERMS_IDENTIFIER
+            self.feed_with_id_and_dcterms_identifier
         )
 
         # First book doesn't have <dcterms:identifier>, so <id> must be used as identifier
-        print(metadata)
         book_1 = metadata.get('https://root.uri/1')
         assert_not_equal(book_1, None)
-        # Seconf book have <id> and <dcterms:identifier>, so <dcters:identifier> must be used as id
+        # Second book have <id> and <dcterms:identifier>, so <dcters:identifier> must be used as id
         book_2 = metadata.get('urn:isbn:9781468316438')
         assert_not_equal(book_2, None)
+        # Verify if id was add in the end of identifier
+        book_2_identifiers = book_2.identifiers
+        found = False
+        for entry in book_2.identifiers:
+            if entry.identifier == 'https://root.uri/2':
+                found = True
+                break
+        eq_(found, True)
+        # Third book has more than one dcterms:identifers, all of then must be present as metadata identifier
+        book_3 = metadata.get('urn:isbn:9781683351993')
+        assert_not_equal(book_2, None)
+        # Verify if id was add in the end of identifier
+        book_3_identifiers = book_3.identifiers
+        expected_identifier = [
+            '9781683351993',
+            'https://root.uri/3',
+            '9781683351504',
+            '9780312939458',
+        ]
+        for target_identifier in expected_identifier:
+            found = False
+            for entry in book_3.identifiers:
+                if entry.identifier == target_identifier:
+                    found = True
+                    break
+            ok_(found, "Cannot find %s" % target_identifier)
```

Comment:
This is fine, but for next time, make two lists and compare them directly, rather than writing a test that has nested loops -- it's easier to read. Since the ordering of the lists isn't relevant, convert them to sets and compare the sets. So something like this:

```
eq_(
  set([x.identifier for x in book_3.identifiers]),
  set(expected_identifier)
)
```

If there's a failure the error message will make clear the difference between the expected result and the actual.

See this example from elsewhere in test_opds_import.py:
```
        # The pools have presentation editions.
        eq_(set(["The Green Mouse", "Johnny Crow's Party"]),
            set([x.presentation_edition.title for x in pools_g]))
```

Reply:
This would also verify that no unexpected identifiers snuck into `book_3.identifiers`. That's not a huge concern here, but this test only verifies that `book_3.identifiers` contains every expected identifier, not that the expected and actual group is identical.